tasks/install_agent_shell.sh

```diff
@@ -0,0 +1,690 @@
+  run_cmd "/usr/lib/apt/apt-helper download-file '$1' '$2'" 2>$tmp_stderr
+  rc=$?
+
+  # check for 404
+  grep "E: Failed to fetch .* 404 " $tmp_stderr 2>&1 >/dev/null
+  if test $? -eq 0; then
+    critical "ERROR 404"
+    unable_to_retrieve_package
+  fi
+
+  # check for bad return status or empty output
+  if test $rc -ne 0 && test ! -s "$2" ; then
+    capture_tmp_stderr "apthelper"
+    return 1
+  fi
+
+  return 0
+}
+
+do_python3_urllib() {
+  info "Trying python3 (urllib.request)..."
+  run_cmd "python3 -c 'import urllib.request ; urllib.request.urlretrieve(\"$1\", \"$2\")'" 2>$tmp_stderr
+  rc=$?
+
+  # check for 404
+  if grep "404: Not Found" $tmp_stderr 2>&1 >/dev/null ; then
+    critical "ERROR 404"
+    unable_to_retrieve_package
+  fi
+
+  if test $rc -eq 0 && test -s "$2" ; then
+    return 0
+  fi
+
+  capture_tmp_stderr "perl"
+  return 1
+}
+
+# do_perl_lwp URL FILENAME
+do_perl_lwp() {
+  info "Trying perl (LWP::Simple)..."
+  run_cmd "perl -e 'use LWP::Simple; getprint(\$ARGV[0]);' '$1' > '$2' 2>$tmp_stderr"
+  rc=$?
+
+  # check for 404
+  grep "404 Not Found" $tmp_stderr 2>&1 >/dev/null
+  if test $? -eq 0; then
+    critical "ERROR 404"
+    unable_to_retrieve_package
+  fi
+
+  if test $rc -eq 0 && test -s "$2" ; then
+    return 0
+  fi
+
+  capture_tmp_stderr "perl"
+  return 1
+}
+
+# do_perl_ff URL FILENAME
+do_perl_ff() {
+  info "Trying perl (File::Fetch)..."
+  run_cmd "perl -e 'use File::Fetch; use File::Copy; my \$ff = File::Fetch->new(uri => \$ARGV[0]); my \$outfile = \$ff->fetch() or die \$ff->server; copy(\$outfile, \$ARGV[1]) or die \"copy failed: \$!\"; unlink(\$outfile) or die \"delete failed: \$!\";' '$1' '$2' 2>>$tmp_stderr"
+  rc=$?
+
+  # check for 404
+  grep "HTTP response: 404" $tmp_stderr 2>&1 >/dev/null
+  if test $? -eq 0 ; then
+    critical "ERROR 404"
+    unable_to_retrieve_package
+  fi
+
+  if test $rc -eq 0 && test -s "$2" ; then
+    return 0
+  fi
+
+  capture_tmp_stderr "perl"
+  return 1
+}
+
+# do_download URL FILENAME
+do_download() {
+  info "Downloading $1"
+  info "  to file $2"
+
+  # we try all of these until we get success.
+  # perl, in particular may be present but LWP::Simple may not be installed
+
+  if exists wget; then
+    do_wget $1 $2 && return 0
+  fi
+
+  if exists curl; then
+    do_curl $1 $2 && return 0
+  fi
+
+  if exists fetch; then
+    do_fetch $1 $2 && return 0
+  fi
+
+  if exists_perl_lwp; then
+    do_perl_lwp $1 $2 && return 0
+  fi
+
+  if exists_perl_ff; then
+    do_perl_ff $1 $2 && return 0
+  fi
+
+  if exists_python3_urllib; then
+    do_python3_urllib $1 $2 && return 0
+  fi
+
+  if exists_apt_helper; then
+    do_apt_helper $1 $2 && return 0
+  fi
+
+  critical "Cannot download package as none of wget/curl/fetch/perl-LWP-Simple/perl-File-Fetch/python3/apt-helper is found"
+  unable_to_retrieve_package
+}
+
+#####
+# Source options
+#
+
+if [ -n "$PT_yum_source" ]; then
+  yum_source=$PT_yum_source
+else
+  if [ "$nightly" = true ]; then
```

Comment:
Isn't `elif` also available in POSIX sh?
```suggestion
elif [ "$nightly" = true ]; then
```

Reply:
Is this relevant if a bash is required?

Reply:
I thought you were going for maximum compatibility. If you can rely on bash, I'd certainly use the shorter `elif`.

And if you require bash, I'd encourage bash checks.
```suggestion
elif [[ $nightly == true ]]; then
```

This is safer, in case there is some weird garbage in the variable.

Reply:
But it seems to be
https://queirozf.com/entries/posix-shell-tests-and-conditionals-examples-and-reference

Reply:
maximum compatibility would be fine. But do you think there will be a platform without bash for the OpenVox agent?

Reply:
But what would be the behavior if no bash exists on a platform because line 1 is '/usr/bin/env bash'?

Reply:
I removed parameter nightly since OpenVox offers such builds.